includes/compat/subscriptions/class-wc-payment-gateway-wcpay-subscriptions-compat.php

```diff
@@ -136,4 +137,44 @@ public function add_token_to_order( $order, $token ) {
 			parent::add_token_to_order( $subscription, $token );
 		}
 	}
+
+	/**
+	 * Render the payment method used for a subscription in My Account pages
+	 *
+	 * @param string          $payment_method_to_display Default payment method to display.
+	 * @param WC_Subscription $subscription              Subscription object.
+	 *
+	 * @return string Payment method string to display in UI.
+	 */
+	public function maybe_render_subscription_payment_method( $payment_method_to_display, $subscription ) {
+		try {
+			if ( $subscription->get_payment_method() !== $this->id ) {
+				return $payment_method_to_display;
+			}
+
+			$token = $this->get_payment_token( $subscription );
+
+			if ( is_null( $token ) ) {
+				Logger::info( 'There is no saved payment token for subscription #' . $subscription->get_id() );
+				return $payment_method_to_display;
+			}
+			return $token->get_display_name();
+		} catch ( \Exception $e ) {
+			Logger::error( 'Failed to get payment method for subscription  #' . $subscription->get_id() . ' ' . $e );
+			return $payment_method_to_display;
+		}
+	}
+
+	/**
+	 * Retrieve payment token from a subscription or order.
+	 *
+	 * @param WC_Order $order Order or subscription object.
+	 *
+	 * @return null|WC_Payment_Token Last token associated with order or subscription.
+	 */
+	private function get_payment_token( $order ) {
+		$order_tokens = $order->get_payment_tokens();
+		$token_id     = end( $order_tokens );
+		return ! $token_id ? null : WC_Payment_Tokens::get( $token_id );
+	}
```

Comment:
Thank you for refactoring this! Looks great! Do you think it might be worth it to use it in the [`add_token_to_order`](https://github.com/Automattic/woocommerce-payments/blob/master/includes/class-wc-payment-gateway-wcpay.php#L638) logic as well?

Reply:
Good catch! Moved the extracted method to WC_Payment_Gateway_WCPay class and refactored `add_token_to_order` in https://github.com/Automattic/woocommerce-payments/pull/893/commits/ce5c0d44899c99c79b229527e60204bd4aeb9f5d

Also refactored `update_failing_payment_method` in https://github.com/Automattic/woocommerce-payments/pull/893/commits/c49fa38f6a1fe7e125d867d48c68ffaaab423f55

Reply:
> Also refactored update_failing_payment_method in c49fa38

Great catch! 🚀 🎉